File: 0678_Valid Parenthesis String/ValidParenthesisString_BruteForce_TLE.cs
public class Solution {
    private bool isValid = false;
    public bool CheckValidString(string s) {
        var chs = s.ToCharArray();
        Solve(chs, 0);
        return isValid;
    }
    
    private void Solve(char[] chs, int index) {
        if(index == chs.Length) {
            isValid = Valid(chs);
            return;
        }
        
        if(chs[index] == '*') {
            foreach(var c in "() ") {
                chs[index] = c;
                Solve(chs, index+1);
                if(isValid) return;
            }
            chs[index] = '*';
        }else {
            Solve(chs, index+1);
        }
    }
    
    private bool Valid(char[] chs)
    {
        int bal = 0;
        for(int i=0;i<chs.Length;i++) {
            if(chs[i] == '(') bal++;
            if(chs[i] == ')') bal--;
            if(bal < 0) break;
        }
        
        return bal == 0;
    }
}

File: 0678_Valid Parenthesis String/ValidParenthesisString_BruteForce_TLE.java
class Solution {
    public boolean checkValidString(String s) {
        return DFS(s.toCharArray(), 0);
    }
    
    private boolean DFS(char[] chars, int index) {
        if(index == chars.length) {
            if(isValid(chars)) return true;
            return false;
        }
        
        char c = chars[index];

        if(c == '*') {
            for(char ch : "() ".toCharArray()) {
                chars[index] = ch;
                if(DFS(chars, index+1)) return true;
            }

            chars[index] = '*';
        }else {
            if(DFS(chars, index+1)) return true;
        }
        
        return false;
    }
    
    private boolean isValid(char[] chars) {
        int bal = 0;
        for(char c : chars) {
            if(c == '(') {
                bal++;
            }else if(c == ')') {
                if(bal == 0) return false;
                bal--;
            }
        }
        
        return bal == 0;
    }
}